File: common/mrtjp/projectred/core/CoreRecipes.java
package mrtjp.projectred.core;

import mrtjp.projectred.ProjectRedCore;
import mrtjp.projectred.core.ItemPart.EnumPart;
import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.FurnaceRecipes;
import net.minecraftforge.oredict.OreDictionary;
import net.minecraftforge.oredict.ShapedOreRecipe;
import net.minecraftforge.oredict.ShapelessOreRecipe;
import codechicken.microblock.ItemMicroPart;
import codechicken.microblock.handler.MicroblockProxy;
import cpw.mods.fml.common.registry.GameRegistry;

public class CoreRecipes {
    public static void initCoreRecipes() {
        initPartRecipes();
        initToolRecipes();
    }
    
    private static void initToolRecipes() {
        /** Draw Plate **/
        GameRegistry.addRecipe(new ShapedOreNBTRecipe(
                new ItemStack(ProjectRedCore.itemDrawPlate, 1),
                " i ",
                "idi",
                " i ",
                'i', ItemMicroPart.create(770, Block.blockIron.getUnlocalizedName()),
                'd', ItemMicroPart.create(2, Block.blockDiamond.getUnlocalizedName())
        ).setCheckNBT());  
        
        /** Screw Driver **/
        GameRegistry.addRecipe(new ShapedOreRecipe(new ItemStack(ProjectRedCore.itemScrewdriver),
                "i  ",
                " ib",
                " bi",
                'i', Item.ingotIron,
                'b', PRColors.BLUE.getOreDict()
        ));
        
        /** Wire debugger **/
        GameRegistry.addRecipe(new ItemStack(ProjectRedCore.itemWireDebugger), 
                "a a",
                "ber",
                "bgr",
                'a', EnumPart.REDINGOT.getItemStack(),
                'b', new ItemStack(Item.dyePowder, 1, PRColors.BLACK.dyeId()),
                'e', Item.emerald,
                'l', new ItemStack(Item.dyePowder, 1, PRColors.BLUE.dyeId()),
                'r', new ItemStack(Item.dyePowder, 1, PRColors.RED.dyeId()),
                'g', Item.glowstone
        );

    }

    private static void initPartRecipes() {
        /** Circuit Plate **/
        GameRegistry.addSmelting(Block.stone.blockID, EnumPart.PLATE.getItemStack(2), 0f);
        
        /** Conductive Plate **/
        GameRegistry.addRecipe(EnumPart.CONDUCTIVEPLATE.getItemStack(), 
                "r",
                "p",
                'r', Item.redstone,
                'p', EnumPart.PLATE.getItemStack()
        );
                
        /** Anode **/
        GameRegistry.addRecipe(EnumPart.ANODE.getItemStack(3), 
                " r ",
                "rrr",
                "ppp",
                'r', Item.redstone,
                'p', EnumPart.PLATE.getItemStack()
        );
        
        /** Cathode **/
        GameRegistry.addRecipe(EnumPart.CATHODE.getItemStack(), 
                "t",
                "p",
                't', Block.torchRedstoneActive,
                'p', EnumPart.PLATE.getItemStack()
        );
        
        /** Pointer **/
        GameRegistry.addRecipe(EnumPart.POINTER.getItemStack(), 
                "b",
                "m",
                "c",
                'b', Block.stone,
                'm', Block.torchRedstoneActive,
                'c', EnumPart.PLATE.getItemStack()
        );

        /** Silicon Chip **/
        GameRegistry.addRecipe(EnumPart.SILICONCHIP.getItemStack(), 
                " s ",
                "ppp",
                's', EnumPart.INFUSEDSILICON.getItemStack(),
                'p', EnumPart.PLATE.getItemStack()
        );
        
        /** Energized Silicon Chip **/
        GameRegistry.addRecipe(EnumPart.ENERGIZEDSILICONCHIP.getItemStack(), 
                " e ",
                "ppp",
                'e', EnumPart.ENERGIZEDSILICON.getItemStack(),
                'p', EnumPart.PLATE.getItemStack()
        );
        
        /** Platformed Plate **/
        GameRegistry.addRecipe(EnumPart.PLATFORMEDPLATE.getItemStack(), 
                " r ",
                "sps",
                "prp",
                'r', EnumPart.WIREDPLATE.getItemStack(),
                's', Item.stick,
                'p', EnumPart.PLATE.getItemStack()
        );

        
        /** Silicon Boule **/
        FurnaceRecipes.smelting().addSmelting(ProjectRedCore.itemComponent.itemID, EnumPart.SANDYCOALCOMPOUND.meta, EnumPart.SILICONBOULE.getItemStack(), 0);
        
        /** Silicon **/
        GameRegistry.addRecipe(EnumPart.SILICON.getItemStack(8), 
                "s",
                "b",
                's', new ItemStack(MicroblockProxy.sawDiamond(), 1, OreDictionary.WILDCARD_VALUE), 
                'b', EnumPart.SILICONBOULE.getItemStack()
        );
        
        /** Infused Silicon **/
        FurnaceRecipes.smelting().addSmelting(ProjectRedCore.itemComponent.itemID, EnumPart.REDSILICONCOMPOUND.meta, EnumPart.INFUSEDSILICON.getItemStack(), 0);
        
        /** Energized Silicon **/
        FurnaceRecipes.smelting().addSmelting(ProjectRedCore.itemComponent.itemID, EnumPart.GLOWINGSILICONCOMPOUND.meta, EnumPart.ENERGIZEDSILICON.getItemStack(), 0);
        
        /** Motor **/
        GameRegistry.addRecipe(EnumPart.MOTOR.getItemStack(), 
                " i ",
                "scs",
                "rcr",
                'i', Item.ingotIron,
                's', Block.stone,
                'c', EnumPart.COPPERCOIL.getItemStack(),
                'r', Item.redstone
        );
        
        /** Copper Coil **/
        GameRegistry.addRecipe(new ShapedOreRecipe(EnumPart.COPPERCOIL.getItemStack(), 
                "cd",
                'c', "ingotCopper",
                'd', new ItemStack(ProjectRedCore.itemDrawPlate, 1, OreDictionary.WILDCARD_VALUE)
        ));
        
        /** Iron Coil **/
        GameRegistry.addRecipe(EnumPart.IRONCOIL.getItemStack(), 
                "cd",
                'c', new ItemStack(Item.ingotIron),
                'd', new ItemStack(ProjectRedCore.itemDrawPlate, 1, OreDictionary.WILDCARD_VALUE)
        );
        
        /** Gold Coil **/
        GameRegistry.addRecipe(EnumPart.GOLDCOIL.getItemStack(), 
                "cd",
                'c', new ItemStack(Item.ingotGold),
                'd', new ItemStack(ProjectRedCore.itemDrawPlate, 1, OreDictionary.WILDCARD_VALUE)
        );
        
        /** Red Alloy Ingot **/
        FurnaceRecipes.smelting().addSmelting(ProjectRedCore.itemComponent.itemID, EnumPart.REDIRONCOMPOUND.meta, EnumPart.REDINGOT.getItemStack(), 0);
        
        /** Illumar **/
        for (int i = 0; i < EnumPart.ILLUMAR_PARTS.length; i++) {
            EnumPart p = EnumPart.ILLUMAR_PARTS[i];
            GameRegistry.addRecipe(new ShapelessOreRecipe(p.getItemStack(), 
                    new ItemStack(Item.glowstone),
                    new ItemStack(Item.glowstone),
                    PRColors.get(i).getOreDict(),
                    PRColors.get(i).getOreDict()
            ));
        }
        
        /** Woven Cloth **/
        GameRegistry.addRecipe(EnumPart.WOVENCLOTH.getItemStack(), 
                "sss",
                "sws",
                "sss",
                's', Item.silk,
                'w', Item.stick
        );
        
        /** Sail **/
        GameRegistry.addRecipe(EnumPart.SAIL.getItemStack(), 
                "ss",
                "ss",
                "ss",
                's', EnumPart.WOVENCLOTH.getItemStack()
        );
        
        /** Red Iron Compound **/
        GameRegistry.addRecipe(EnumPart.REDIRONCOMPOUND.getItemStack(), 
                "rrr",
                "rir",
                "rrr",
                'r', Item.redstone,
                'i', Item.ingotIron
        );

        /** Sandy Coal Compound **/
        GameRegistry.addRecipe(EnumPart.SANDYCOALCOMPOUND.getItemStack(), 
                "sss",
                "scs",
                "sss",
                'c', Block.coalBlock,
                's', Block.sand
        );
        
        /** Red Silicon Compound **/
        GameRegistry.addRecipe(EnumPart.REDSILICONCOMPOUND.getItemStack(), 
                "rrr",
                "rsr",
                "rrr",
                'r', Item.redstone,
                's', EnumPart.SILICON.getItemStack()
        );

        /** Glowing Silicon Compound **/
        GameRegistry.addRecipe(EnumPart.GLOWINGSILICONCOMPOUND.getItemStack(), 
                "ggg",
                "gsg",
                "ggg",
                'g', Item.glowstone,
                's', EnumPart.SILICON.getItemStack()
        );

    }

}

File: archive/scala/mrtjp/projectred/archive/CoreRecipes.scala
package mrtjp.projectred.archive

import java.lang.{Character => JChar}

import codechicken.lib.colour.EnumColour
import codechicken.microblock.handler.MicroblockProxy
import com.google.gson.JsonObject
import mrtjp.projectred.ProjectRedCore
import mrtjp.projectred.core.PartDefs
import mrtjp.projectred.core.PartDefs._
import Conversions._
import net.minecraft.block.Block
import net.minecraft.init.{Blocks, Items}
import net.minecraft.item.ItemStack
import net.minecraftforge.oredict.OreDictionary

object CoreRecipes {

    var dumper: RecipeDumper = _

    def initCoreRecipes() {
        dumper = new RecipeDumper("core")
        initPartRecipes()
        initToolRecipes()
        dumper.dump
    }

    private def initToolRecipes() {
        def p(fac: String, size: Int, block: Block): JsonObject = {
            new JsonObject().add("type", "microblockcbe:micro_material").add("factory", fac).add("size", size).add_("material", new JsonObject().add("block", block.getRegistryName.toString))
        }

        /** Draw Plate **/
        dumper.addRecipe(new ShapedOreRecipe(new ItemStack(ProjectRedCore.itemDrawPlate),
            " i ", "idi", " i ",
            'i': JChar, p("edge", 2, Blocks.IRON_BLOCK),
            'd': JChar, p("face", 2, Blocks.DIAMOND_BLOCK)
        )).setJsonName("tools\\draw_plate")

        /** Panel Reset recipe **/
        dumper.addRecipe(new ShapelessOreRecipe(new ItemStack(Items.DIAMOND, 2),
            p("face", 2, Blocks.DIAMOND_BLOCK)
        )).setJsonName("misc\\panel_2_diamonds")

        /** Screwdriver **/
        dumper.addRecipe(new ShapedOreRecipe(new ItemStack(ProjectRedCore.itemScrewdriver),
            "i  ", " ib", " bi",
            'i': JChar, "ingotIron",
            'b': JChar, "dyeBlue"
        )).setJsonName("tools\\screwdriver")

        /** Wire Debugger **/
        dumper.addRecipe(new ShapedOreRecipe(new ItemStack(ProjectRedCore.itemMultimeter),
            "a a", "ber", "bgr",
            'a': JChar, "ingotRedAlloy",
            'b': JChar, "dyeBlack",
            'e': JChar, "dyeGreen",
            'r': JChar, "dyeRed",
            'g': JChar, "dustGlowstone"
        )).setJsonName("tools\\multimeter")
    }

    private def initPartRecipes() {
        /** Conductive Plate **/
        dumper.addRecipe(new ShapedOreRecipe(CONDUCTIVEPLATE.makeStack,
            "r", "p",
            'r': JChar, "dustRedstone",
            'p': JChar, PLATE.makeStack
        )).setJsonName("parts\\conductive_plate")

        /** Anode **/
        dumper.addRecipe(new ShapedOreRecipe(ANODE.makeStack(3),
            " r ", "rrr", "ppp",
            'r': JChar, "dustRedstone",
            'p': JChar, PLATE.makeStack
        )).setJsonName("parts\\anode")

        /** Cathode **/
        dumper.addRecipe(CATHODE.makeStack,
            "t", "p",
            't': JChar, Blocks.REDSTONE_TORCH,
            'p': JChar, PLATE.makeStack
        ).setJsonName("parts\\cathode")

        /** Pointer **/
        dumper.addRecipe(new ShapedOreRecipe(POINTER.makeStack,
            "b", "m", "c",
            'b': JChar, "stone",
            'm': JChar, Blocks.REDSTONE_TORCH,
            'c': JChar, PLATE.makeStack
        )).setJsonName("parts\\pointer")

        /** Silicon Chip **/
        dumper.addRecipe(SILICONCHIP.makeStack,
            " s ", "ppp", 's': JChar,
            INFUSEDSILICON.makeStack,
            'p': JChar, PLATE.makeStack
        ).setJsonName("parts\\silicon_chip")

        /** Energized Silicon Chip **/
        dumper.addRecipe(ENERGIZEDSILICONCHIP.makeStack,
            " e ", "ppp", 'e': JChar,
            ENERGIZEDSILICON.makeStack,
            'p': JChar, PLATE.makeStack
        ).setJsonName("parts\\energized_silicon_chip")

        /** Platformed Plate **/
        dumper.addRecipe(new ShapedOreRecipe(PLATFORMEDPLATE.makeStack,
            " r ", "sps", "prp",
            'r': JChar, WIREDPLATE.makeStack,
            's': JChar, "stickWood",
            'p': JChar, PLATE.makeStack)
        ).setJsonName("parts\\platformed_plate")


        /** Silicon **/
        dumper.addRecipe(SILICON.makeStack(8),
            "s", "b",
            's': JChar, new ItemStack(MicroblockProxy.sawDiamond, 1, OreDictionary.WILDCARD_VALUE),
            'b': JChar, SILICONBOULE.makeStack
        ).setJsonName("resource\\silicon")


        /** Motor **/
        dumper.addRecipe(new ShapedOreRecipe(MOTOR.makeStack,
            " i ", "scs", "rcr",
            'i': JChar, "ingotIron",
            's': JChar, "stone", 'c': JChar, COPPERCOIL.makeStack,
            'r': JChar, "dustRedstone")
        ).setJsonName("misc\\motor")

        /** Copper Coil **/
        dumper.addRecipe(new ShapedOreRecipe(COPPERCOIL.makeStack,
            "cd",
            'c': JChar, "ingotCopper",
            'd': JChar, new ItemStack(ProjectRedCore.itemDrawPlate, 1, OreDictionary.WILDCARD_VALUE))
        ).setJsonName("misc\\copper_coil")

        /** Iron Coil **/
        dumper.addRecipe(new ShapedOreRecipe(IRONCOIL.makeStack,
            "cd",
            'c': JChar, "ingotIron",
            'd': JChar, new ItemStack(ProjectRedCore.itemDrawPlate, 1, OreDictionary.WILDCARD_VALUE))
        ).setJsonName("misc\\iron_coil")

        /** Gold Coil **/
        dumper.addRecipe(new ShapedOreRecipe(GOLDCOIL.makeStack,
            "cd",
            'c': JChar, "ingotGold",
            'd': JChar, new ItemStack(ProjectRedCore.itemDrawPlate, 1, OreDictionary.WILDCARD_VALUE))
        ).setJsonName("misc\\gold_coil")


        //TODO
        /** Illumar **/
        for (i <- 0 until 16) {
            val p: PartDefs.PartVal = PartDefs.ILLUMARS(i)
            dumper.addRecipe(new ShapelessOreRecipe(p.makeStack, "dustGlowstone", "dustGlowstone",
                EnumColour.values.apply(i).getDyeOreName, EnumColour.values.apply(i).getDyeOreName).setJsonName(s"lumar\\lumar_${EnumColour.values.apply(i).getName}"))
        }

        /** Woven Cloth **/
        dumper.addRecipe(new ShapedOreRecipe(WOVENCLOTH.makeStack,
            "sss", "sws", "sss",
            's': JChar, "string",
            'w': JChar, "stickWood")
        ).setJsonName("misc\\woven_cloth")

        /** Sail **/
        dumper.addRecipe(SAIL.makeStack,
            "ss", "ss", "ss",
            's': JChar, WOVENCLOTH.makeStack
        ).setJsonName("misc\\sail")

        /** Red Iron Compound **/
        dumper.addRecipe(new ShapedOreRecipe(REDIRONCOMPOUND.makeStack,
            "rrr", "rir", "rrr",
            'r': JChar, "dustRedstone",
            'i': JChar, "ingotIron")
        ).setJsonName("resource\\red_iron_compound")

        /** Sandy Coal Compound **/
        dumper.addRecipe(new ShapedOreRecipe(SANDYCOALCOMPOUND.makeStack,
            "sss", "scs", "sss",
            'c': JChar, "blockCoal",
            's': JChar, "sand")
        ).setJsonName("resource\\sandy_coal_compound")

        /** Red Silicon Compound **/
        dumper.addRecipe(new ShapedOreRecipe(REDSILICONCOMPOUND.makeStack,
            "rrr", "rsr", "rrr",
            'r': JChar, "dustRedstone",
            's': JChar, SILICON.makeStack)
        ).setJsonName("resource\\red_silicon_compound")

        /** Glowing Silicon Compound **/
        dumper.addRecipe(new ShapedOreRecipe(GLOWINGSILICONCOMPOUND.makeStack,
            "ggg", "gsg", "ggg",
            'g': JChar, "dustGlowstone",
            's': JChar, SILICON.makeStack)
        ).setJsonName("resource\\glowing_silicon_compound")

        /** Electrotine Iron Compound **/
        dumper.addRecipe(new ShapedOreRecipe(ELECTROTINEIRONCOMPOUND.makeStack,
            "bbb", "bib", "bbb",
            'b': JChar, "dustElectrotine",
            'i': JChar, "ingotIron")
        ).setJsonName("resource\\electrotine_iron_compound")

        /** Electrotine Silicon Compound **/
        dumper.addRecipe(new ShapedOreRecipe(ELECTROTINESILICONCOMPOUND.makeStack,
            "bbb", "bsb", "bbb",
            'b': JChar, "dustElectrotine",
            's': JChar, SILICON.makeStack)
        ).setJsonName("resource\\electrotine_silicon_compound")
    }
}
